File: src/oop/interfaces/java/Interface.java
package oop.interfaces.java;

/*
An interface in Java is a blueprint of a class. It has static constants and abstract methods.

The interface in Java is a mechanism to achieve abstraction. There can be only abstract methods in the Java interface, not method body.
It is used to achieve abstraction and multiple inheritance in Java.

In other words, you can say that interfaces can have abstract methods and variables. It cannot have a method body.

Java Interface also represents the IS-A relationship.

It cannot be instantiated just like the abstract class.

Since Java 8, we can have default and static methods in an interface.

Since Java 9, we can have private methods in an interface.
 */

public class Interface {

    //Interface declaration: by first user
    interface Drawable {
        void draw();
    }

    //Implementation: by second user
    class Rectangle implements Drawable {
        public void draw() {
            System.out.println("drawing rectangle");
        }
    }

    static class Circle implements Drawable {
        public void draw() {
            System.out.println("drawing circle");
        }
    }
    //Using interface: by third user

    public static void main(String args[]) {
        Interface.Drawable d = new Interface.Circle();//In real scenario, object is provided by method e.g. getDrawable()
        d.draw();
    }

}

File: src/oop/interfaces/kotlin/Interface.kt
package oop.interfaces.kotlin

/*
Interfaces are custom types provided by Kotlin that cannot be instantiated directly. Instead, these define a form of behavior that the implementing types have to follow.
With the interface, you can define a set of properties and methods, that the concrete types must follow and implement.
 */
class Interface {

    interface Vehicle {
        fun start() // abstract method
        fun stop()
    }

    class Car : Vehicle {
        override fun start() {
            println("Car started")
        }

        override fun stop() {
            println("Car stopped")
        }
    }


}

fun main() {
    val obj = Interface.Car()
    obj.start()
    obj.stop()
}